File: skills/species-distribution-modeling/scripts/predict_distribution.R
# ecological-agent-skills / Copyright (C) 2026 Francisco Diego Barros Barata
# SPDX-License-Identifier: GPL-3.0-or-later

# Usage: Rscript predict_distribution.R <model_rds> <predictor_stack_tif> <output_dir> [threshold_method] [scenario_label]
# Predict suitability from a fitted SDM (maxnet, gbm, randomForest, or ensemble list).
# Applies MaxTSS / P10 / MTP thresholding, computes MESS, and saves all rasters.

# ── Inline logger ─────────────────────────────────────────────────────────────
SKILL_NAME <- "species-distribution-modeling"
.log_ts  <- function() format(Sys.time(), "[%Y-%m-%d %H:%M:%S]")
log_info <- function(...) message(.log_ts(), " [INFO]  ", sprintf(...))
log_warn <- function(...) message(.log_ts(), " [WARN]  ", sprintf(...))
log_error<- function(...) message(.log_ts(), " [ERROR] ", sprintf(...))
log_step <- function(n, d) log_info("-- STEP %d: %s", n, d)
log_decision <- function(v, val, why) log_info("DECISION | %s = %s | %s", v, val, why)
dir.create("logs", recursive = TRUE, showWarnings = FALSE)

suppressPackageStartupMessages({
  library(terra)
  library(dismo)
})

# ── Arguments ─────────────────────────────────────────────────────────────────
args             <- commandArgs(trailingOnly = TRUE)
model_rds        <- if (length(args) >= 1) args[1] else stop("model_rds required")
predictor_tif    <- if (length(args) >= 2) args[2] else stop("predictor_stack_tif required")
output_dir       <- if (length(args) >= 3) args[3] else "outputs/predictions"
threshold_method <- if (length(args) >= 4) args[4] else "MaxTSS"
scenario_label   <- if (length(args) >= 5) args[5] else "current"

log_decision("threshold_method", threshold_method,
             "MaxTSS balances sensitivity and specificity; P10 is more conservative; MTP is permissive")
log_decision("scenario_label", scenario_label, "Label embedded in output filenames")

# ── Precondition checks ───────────────────────────────────────────────────────
if (!file.exists(model_rds)) {
  log_error("Model RDS nao encontrado: %s\nCausa provavel: run_ensemble_sdm.R nao foi executado.\nVerifique: a saida de skills/species-distribution-modeling.\nSkill anterior: species-distribution-modeling", model_rds)
  stop("Missing model file: ", model_rds)
}
if (!file.exists(predictor_tif)) {
  log_error("Predictor stack nao encontrado: %s\nCausa provavel: stack nao foi preparado.\nVerifique: geoprocessing-for-ecology / prepare_future_layers.R.\nSkill anterior: geoprocessing-for-ecology", predictor_tif)
  stop("Missing predictor stack: ", predictor_tif)
}
dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

log_step(1, "Loading model and predictor stack")
model_obj  <- tryCatch(readRDS(model_rds), error = function(e) {
  log_error("Falha ao ler model RDS: %s\nCausa provavel: arquivo corrompido ou formato incompativel.\nVerifique: Rscript que gerou o modelo usou saveRDS().", conditionMessage(e))
  stop(e)
})
predictors <- tryCatch(rast(predictor_tif), error = function(e) {
  log_error("Falha ao ler raster stack: %s\nCausa provavel: arquivo GeoTIFF invalido ou caminho errado.\nVerifique: predictor_tif existe e tem multiplas camadas.", conditionMessage(e))
  stop(e)
})
log_info("Model loaded: class = %s", class(model_obj)[1])
log_info("Predictor stack: %d layers, CRS = %s", nlyr(predictors), crs(predictors, describe = TRUE)$name)

# Detect ensemble (list of models) vs single model
is_ensemble <- is.list(model_obj) && !inherits(model_obj, "MaxEnt")
log_decision("is_ensemble", is_ensemble,
             "Ensemble if model_rds is a named list with $models and $auc_weights")

# ── Step 2: Reproject predictors if needed ────────────────────────────────────
log_step(2, "Checking and aligning CRS / resolution")
if (is_ensemble) {
  ref_model <- model_obj$models[[1]]
} else {
  ref_model <- model_obj
}

# ── Step 3: Generate suitability predictions ──────────────────────────────────
log_step(3, "Predicting suitability")

predict_single <- function(mod, preds) {
  cls <- class(mod)[1]
  log_info("Predicting with model class: %s", cls)
  if (cls == "maxnet") {
    suppressPackageStartupMessages(library(maxnet))
    pred_df <- as.data.frame(preds, na.rm = FALSE)
    p        <- predict(mod, pred_df, type = "cloglog")
    r        <- preds[[1]]; values(r) <- p
    return(r)
  } else if (cls %in% c("gbm", "BRT")) {
    suppressPackageStartupMessages(library(gbm))
    pred_df <- as.data.frame(preds, na.rm = FALSE)
    p        <- predict.gbm(mod, pred_df, n.trees = mod$n.trees, type = "response")
    r        <- preds[[1]]; values(r) <- p
    return(r)
  } else if (cls == "randomForest") {
    suppressPackageStartupMessages(library(randomForest))
    pred_df <- as.data.frame(preds, na.rm = FALSE)
    p        <- predict(mod, pred_df, type = "prob")[, 2]
    r        <- preds[[1]]; values(r) <- p
    return(r)
  } else {
    # Generic predict (MaxEnt, etc.)
    tryCatch(
      predict(mod, predictors),
      error = function(e) {
        log_error("Falha na predicao generica: %s\nCausa provavel: model class nao suportada.\nVerifique: model_obj e um dos tipos suportados (maxnet, gbm, randomForest).", conditionMessage(e))
        stop(e)
      }
    )
  }
}

if (is_ensemble) {
  log_info("Ensemble prediction: %d models", length(model_obj$models))
  suit_stack <- tryCatch({
    preds_list <- lapply(model_obj$models, predict_single, preds = predictors)
    rast(preds_list)
  }, error = function(e) {
    log_error("Falha na predicao ensemble: %s\nCausa provavel: modelos incompativeis.\nVerifique: todos os modelos foram treinados com os mesmos preditores.", conditionMessage(e))
    stop(e)
  })

  weights <- if (!is.null(model_obj$auc_weights)) model_obj$auc_weights else
             rep(1 / nlyr(suit_stack), nlyr(suit_stack))
  log_decision("ensemble_weights", paste(round(weights, 3), collapse = ","),
               "AUC-based weights from model object; uniform if not available")

  suit_mean <- app(suit_stack, fun = function(x) weighted.mean(x, weights, na.rm = TRUE))
  suit_sd   <- app(suit_stack, fun = function(x) sd(x,   na.rm = TRUE))
  log_info("Ensemble uncertainty (mean SD across pixels): %.4f",
           mean(values(suit_sd), na.rm = TRUE))

  unc_file <- file.path(output_dir, paste0("uncertainty_", scenario_label, ".tif"))
  writeRaster(suit_sd, unc_file, overwrite = TRUE)
  log_info("Uncertainty raster saved: %s", unc_file)
} else {
  suit_mean <- predict_single(ref_model, predictors)
}

suit_file <- file.path(output_dir, paste0("suitability_", scenario_label, ".tif"))
writeRaster(suit_mean, suit_file, overwrite = TRUE)
log_info("Suitability raster saved: %s", suit_file)

# ── Step 4: Threshold to binary map ───────────────────────────────────────────
log_step(4, paste("Binarising with threshold method:", threshold_method))

# Look for training presence values (from model object)
threshold_val <- NA_real_
if (!is.null(model_obj$threshold_values)) {
  tv <- model_obj$threshold_values
  threshold_val <- switch(threshold_method,
    MaxTSS = tv["MaxTSS"],
    P10    = tv["P10"],
    MTP    = tv["MTP"],
    {
      log_warn("Threshold method '%s' not found; falling back to MaxTSS", threshold_method)
      tv["MaxTSS"]
    }
  )
} else {
  # Fallback: use quantile of suitability (P10 approximation)
  threshold_val <- quantile(values(suit_mean), 0.10, na.rm = TRUE)
  log_warn("No threshold_values in model object; using 10th percentile of suitability map (%.4f)", threshold_val)
}
log_decision("threshold_value", round(threshold_val, 4),
             paste("Derived via", threshold_method, "from model or suitability distribution"))

binary_map <- suit_mean >= threshold_val
bin_file   <- file.path(output_dir, paste0("binary_", scenario_label, ".tif"))
writeRaster(binary_map, bin_file, overwrite = TRUE, datatype = "INT1U")
log_info("Binary map saved: %s (threshold = %.4f)", bin_file, threshold_val)

# ── Step 5: MESS analysis ──────────────────────────────────────────────────────
log_step(5, "Computing MESS (multivariate environmental similarity surface)")
mess_map <- tryCatch({
  # Use training data ranges stored in model object
  if (!is.null(model_obj$training_ranges)) {
    ranges  <- model_obj$training_ranges
    pred_df <- as.data.frame(predictors, na.rm = FALSE)
    # Simplified MESS: fraction of pixels within training range per variable
    in_range <- mapply(function(col, nm) {
      col >= ranges[nm, "min"] & col <= ranges[nm, "max"]
    }, pred_df, names(pred_df), SIMPLIFY = FALSE)
    in_range_r <- rast(lapply(in_range, function(v) {
      r <- predictors[[1]]; values(r) <- as.integer(v); r
    }))
    app(in_range_r, mean, na.rm = TRUE)  # fraction of vars in range (1 = fully novel = 0)
  } else {
    # Use dismo::mess if available
    suppressPackageStartupMessages(library(dismo))
    if (!is.null(model_obj$training_data)) {
      mess(predictors, model_obj$training_data)
    } else {
      log_warn("No training_ranges or training_data in model; MESS not computed.")
      NULL
    }
  }
}, error = function(e) {
  log_warn("MESS computation failed: %s. Skipping MESS output.", conditionMessage(e))
  NULL
})

if (!is.null(mess_map)) {
  mess_file <- file.path(output_dir, paste0("mess_", scenario_label, ".tif"))
  writeRaster(mess_map, mess_file, overwrite = TRUE)

  pct_novel <- mean(values(mess_map) < 0, na.rm = TRUE) * 100
  if (pct_novel > 20) {
    log_warn("%.1f%% of the prediction area is in novel climate space (MESS < 0). Predictions in those areas are unreliable.", pct_novel)
  } else {
    log_info("MESS: %.1f%% of area is novel climate space (threshold: 20%%)", pct_novel)
  }
  log_info("MESS raster saved: %s", mess_file)
}

# ── Step 6: Summary CSV ────────────────────────────────────────────────────────
log_step(6, "Computing prediction summary statistics")

total_cells    <- sum(!is.na(values(suit_mean)))
suitable_cells <- sum(values(binary_map) == 1, na.rm = TRUE)
cell_area_km2  <- prod(res(suit_mean)) / 1e6  # assumes metres CRS; adjust if degrees
area_total_km2 <- total_cells    * cell_area_km2
area_suit_km2  <- suitable_cells * cell_area_km2

summary_df <- data.frame(
  scenario          = scenario_label,
  threshold_method  = threshold_method,
  threshold_value   = round(threshold_val, 4),
  total_area_km2    = round(area_total_km2, 1),
  suitable_area_km2 = round(area_suit_km2, 1),
  pct_suitable      = round(100 * suitable_cells / total_cells, 2),
  mean_suitability  = round(mean(values(suit_mean), na.rm = TRUE), 4),
  stringsAsFactors  = FALSE
)

sum_file <- file.path(output_dir, "prediction_summary.csv")
write.csv(summary_df, sum_file, row.names = FALSE)
log_info("Summary: suitable area = %.1f km2 (%.1f%% of %.1f km2 total)",
         area_suit_km2, summary_df$pct_suitable, area_total_km2)
log_info("Prediction summary saved: %s", sum_file)

log_step(7, "Done — all outputs written to: %s", output_dir)

File: skills/species-distribution-modeling/scripts/predict_distribution.py
# ecological-agent-skills / Copyright (C) 2026 Francisco Diego Barros Barata
# SPDX-License-Identifier: GPL-3.0-or-later

"""predict_distribution.py — Predict species suitability from a trained sklearn model.

Usage:
    python predict_distribution.py <model_pkl> <predictor_tif> <output_dir>
                                   [--threshold MaxTSS|P10|MTP] [--scenario current]

Outputs:
    suitability_{scenario}.tif   Continuous suitability [0,1]
    binary_{scenario}.tif        Binary presence/absence
    mess_{scenario}.tif          Multivariate environmental similarity
    prediction_summary.csv       Area statistics
"""

import logging
import sys
from datetime import datetime
from pathlib import Path

SKILL_NAME = "species-distribution-modeling"
_LOG_DIR   = Path("logs")
_LOG_DIR.mkdir(parents=True, exist_ok=True)
_log_file  = _LOG_DIR / f"skill_{SKILL_NAME}_{datetime.now().strftime('%Y%m%d_%H%M%S')}.log"
logging.basicConfig(
    level=logging.INFO,
    format="[%(asctime)s] [%(levelname)s] [" + SKILL_NAME + "] %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
    handlers=[
        logging.StreamHandler(sys.stdout),
        logging.FileHandler(_log_file, encoding="utf-8"),
    ],
)
logger = logging.getLogger(SKILL_NAME)


def log_step(n: int, desc: str) -> None:
    logger.info("-- STEP %d: %s", n, desc)


def log_decision(var: str, val, why: str) -> None:
    logger.info("DECISION | %s = %s | %s", var, val, why)


import argparse
import csv
import pickle
import warnings

import numpy as np

try:
    import rasterio
    from rasterio.transform import from_bounds
except ImportError:
    logger.error("rasterio not installed. Run: pip install rasterio")
    sys.exit(1)

# ── Arguments ─────────────────────────────────────────────────────────────────
parser = argparse.ArgumentParser(description="SDM suitability prediction")
parser.add_argument("model_pkl",      help="Path to trained model (.pkl or .joblib)")
parser.add_argument("predictor_tif",  help="Multi-band GeoTIFF predictor stack")
parser.add_argument("output_dir",     help="Directory for outputs")
parser.add_argument("--threshold", default="MaxTSS",
                    choices=["MaxTSS", "P10", "MTP"],
                    help="Threshold method for binary map")
parser.add_argument("--scenario", default="current",
                    help="Label appended to output filenames")
args = parser.parse_args()

log_decision("threshold", args.threshold,
             "MaxTSS balances sensitivity/specificity; P10 conservative; MTP permissive")
log_decision("scenario", args.scenario, "Embedded in output filenames")

# ── Precondition checks ───────────────────────────────────────────────────────
if not Path(args.model_pkl).exists():
    logger.error(
        "Model file not found: %s\n"
        "  Causa provavel: sdm_pipeline.py nao foi executado ou falhou.\n"
        "  Skill anterior que deveria ter produzido este input: species-distribution-modeling",
        args.model_pkl,
    )
    sys.exit(1)

if not Path(args.predictor_tif).exists():
    logger.error(
        "Predictor stack not found: %s\n"
        "  Causa provavel: stack nao preparado.\n"
        "  Skill anterior que deveria ter produzido este input: geoprocessing-for-ecology",
        args.predictor_tif,
    )
    sys.exit(1)

Path(args.output_dir).mkdir(parents=True, exist_ok=True)

# ── Step 1: Load model ────────────────────────────────────────────────────────
log_step(1, "Loading model")
try:
    if args.model_pkl.endswith(".joblib"):
        import joblib
        model_obj = joblib.load(args.model_pkl)
    else:
        with open(args.model_pkl, "rb") as f:
            model_obj = pickle.load(f)
    logger.info("Model loaded: %s", type(model_obj).__name__)
except Exception as e:
    logger.error(
        "Falha ao carregar modelo: %s\n"
        "  Causa provavel: arquivo corrompido ou versao incompativel do sklearn.\n"
        "  Verifique: o modelo foi salvo com a mesma versao do sklearn.",
        e,
    )
    raise

# Support dict with 'model', 'threshold_values', 'training_ranges', 'feature_names'
if isinstance(model_obj, dict):
    model       = model_obj.get("model", model_obj)
    thresh_vals = model_obj.get("threshold_values", {})
    train_ranges= model_obj.get("training_ranges", None)
    feat_names  = model_obj.get("feature_names", None)
else:
    model       = model_obj
    thresh_vals = {}
    train_ranges= None
    feat_names  = None

# ── Step 2: Load predictor stack ──────────────────────────────────────────────
log_step(2, "Loading predictor raster stack")
try:
    with rasterio.open(args.predictor_tif) as src:
        meta        = src.meta.copy()
        band_data   = src.read()            # shape: (n_bands, rows, cols)
        nodata      = src.nodata
        transform   = src.transform
        crs         = src.crs
        band_names  = [src.descriptions[i] or f"band{i+1}" for i in range(src.count)]
    logger.info("Stack: %d bands, shape %s, CRS=%s", band_data.shape[0], band_data.shape[1:], crs)
except Exception as e:
    logger.error("Falha ao ler predictor stack: %s\n  Verifique: arquivo GeoTIFF valido com multiplas bandas.", e)
    raise

n_bands, rows, cols = band_data.shape
X = band_data.reshape(n_bands, -1).T        # (n_pixels, n_bands)

# Mask NoData pixels
if nodata is not None:
    valid_mask = ~np.any(band_data == nodata, axis=0).ravel()
else:
    valid_mask = ~np.any(np.isnan(band_data), axis=0).ravel()

log_decision("valid_pixels", int(valid_mask.sum()), "Pixels without NoData in all bands")

# ── Step 3: Predict suitability ───────────────────────────────────────────────
log_step(3, "Predicting suitability")
suitability = np.full(rows * cols, np.nan, dtype=np.float32)

try:
    X_valid = X[valid_mask]
    if hasattr(model, "predict_proba"):
        proba = model.predict_proba(X_valid)
        # Take probability of class 1 (presence)
        pos_idx = list(model.classes_).index(1) if hasattr(model, "classes_") else 1
        suitability[valid_mask] = proba[:, pos_idx].astype(np.float32)
    elif hasattr(model, "decision_function"):
        # SVM / similar — normalise to [0,1] with sigmoid
        raw = model.decision_function(X_valid)
        suitability[valid_mask] = (1 / (1 + np.exp(-raw))).astype(np.float32)
    else:
        suitability[valid_mask] = model.predict(X_valid).astype(np.float32)
    logger.info("Suitability range: [%.4f, %.4f]",
                float(np.nanmin(suitability)), float(np.nanmax(suitability)))
except Exception as e:
    logger.error("Unexpected error in suitability prediction: %s", e)
    raise

suit_map = suitability.reshape(rows, cols)

# Write suitability raster
suit_meta = meta.copy()
suit_meta.update(count=1, dtype="float32", nodata=-9999.0)
suit_file = Path(args.output_dir) / f"suitability_{args.scenario}.tif"
try:
    with rasterio.open(suit_file, "w", **suit_meta) as dst:
        suit_out = suit_map.copy()
        suit_out[~valid_mask.reshape(rows, cols)] = -9999.0
        dst.write(suit_out[np.newaxis, :, :])
    logger.info("Suitability raster saved: %s", suit_file)
except Exception as e:
    logger.error("Falha ao salvar suitability raster: %s\n  Verifique: permissao de escrita em %s.", e, args.output_dir)
    raise

# ── Step 4: Threshold → binary map ───────────────────────────────────────────
log_step(4, f"Binarising with threshold method: {args.threshold}")
if thresh_vals and args.threshold in thresh_vals:
    thresh = thresh_vals[args.threshold]
else:
    thresh = float(np.nanpercentile(suitability[valid_mask], 10))
    logger.warning("No threshold_values in model; using 10th percentile of suitability (%.4f)", thresh)

log_decision("threshold_value", round(thresh, 4), f"Derived via {args.threshold}")

binary_map = (suit_map >= thresh).astype(np.uint8)
binary_map[~valid_mask.reshape(rows, cols)] = 255   # nodata

bin_meta = meta.copy()
bin_meta.update(count=1, dtype="uint8", nodata=255)
bin_file = Path(args.output_dir) / f"binary_{args.scenario}.tif"
with rasterio.open(bin_file, "w", **bin_meta) as dst:
    dst.write(binary_map[np.newaxis, :, :])
logger.info("Binary map saved: %s (threshold=%.4f)", bin_file, thresh)

# ── Step 5: MESS ──────────────────────────────────────────────────────────────
log_step(5, "Computing MESS")
if train_ranges is not None:
    try:
        mess_vals = np.full(rows * cols, np.nan, dtype=np.float32)
        X_v       = X[valid_mask]
        in_range  = np.ones(len(X_v), dtype=np.float32)
        for j, nm in enumerate(feat_names or band_names):
            if nm in train_ranges:
                lo, hi = train_ranges[nm]["min"], train_ranges[nm]["max"]
                in_range *= ((X_v[:, j] >= lo) & (X_v[:, j] <= hi)).astype(np.float32)
        # in_range==0 → novel; map to negative score for convention
        mess_score = 2 * in_range - 1   # +1 = analogue, -1 = novel
        mess_vals[valid_mask] = mess_score

        mess_map_arr = mess_vals.reshape(rows, cols)
        mess_meta    = meta.copy()
        mess_meta.update(count=1, dtype="float32", nodata=-9999.0)
        mess_file = Path(args.output_dir) / f"mess_{args.scenario}.tif"
        with rasterio.open(mess_file, "w", **mess_meta) as dst:
            dst.write(mess_map_arr[np.newaxis, :, :])

        pct_novel = float(np.mean(mess_score < 0) * 100)
        if pct_novel > 20:
            logger.warning("%.1f%% of prediction area is in novel climate space (MESS < 0). Extrapolation risk is HIGH.", pct_novel)
        else:
            logger.info("MESS: %.1f%% novel climate space", pct_novel)
        logger.info("MESS raster saved: %s", mess_file)
    except Exception as e:
        logger.warning("MESS computation failed: %s. Skipping.", e)
else:
    logger.warning("No training_ranges in model object; MESS not computed.")

# ── Step 6: Summary CSV ───────────────────────────────────────────────────────
log_step(6, "Computing summary statistics")
total_px    = int(valid_mask.sum())
suitable_px = int(np.sum(binary_map[valid_mask.reshape(rows, cols)] == 1))

# Pixel area (approximate: assumes metres if large values, else degrees)
px_size_x = abs(transform.a)
px_size_y = abs(transform.e)
if px_size_x > 1:   # metres
    px_area_km2 = px_size_x * px_size_y / 1e6
else:               # degrees → rough km² at equator
    px_area_km2 = px_size_x * 111 * px_size_y * 111

area_total_km2 = total_px    * px_area_km2
area_suit_km2  = suitable_px * px_area_km2
pct_suit       = 100.0 * suitable_px / total_px if total_px > 0 else 0.0
mean_suit      = float(np.nanmean(suitability[valid_mask]))

summary = {
    "scenario":           args.scenario,
    "threshold_method":   args.threshold,
    "threshold_value":    round(thresh, 4),
    "total_area_km2":     round(area_total_km2, 1),
    "suitable_area_km2":  round(area_suit_km2, 1),
    "pct_suitable":       round(pct_suit, 2),
    "mean_suitability":   round(mean_suit, 4),
}

sum_file = Path(args.output_dir) / "prediction_summary.csv"
with open(sum_file, "w", newline="", encoding="utf-8") as f:
    w = csv.DictWriter(f, fieldnames=list(summary.keys()))
    w.writeheader(); w.writerow(summary)

logger.info("Suitable area: %.1f km2 (%.1f%% of %.1f km2)", area_suit_km2, pct_suit, area_total_km2)
logger.info("Summary saved: %s", sum_file)

log_step(7, "Done — all outputs in: %s", args.output_dir)


if __name__ == "__main__":
    pass   # argparse executed at module level above
